File: assets/Scripts/environment/sky/FogController.js
"use strict";

const SkyMath = require("./SkyMath.js");

class FogController {
    constructor() {
        this.fogBase = { r: 0.70, g: 0.78, b: 0.90 };
        this.fogDistance = 250.0;
        this.fogDensityDay = 1.10;
        this.fogDensityNight = 1.35;

        this._lastFogDensity = NaN;
        this._lastFogColorKey = "";
        this._lastFogDistance = NaN;
    }

    applyCfg(cfg) {
        if (!cfg || !cfg.fog) return;

        const fog = cfg.fog;

        if (fog.color) {
            const c = fog.color;
            if (c.r != null) this.fogBase.r = +c.r;
            if (c.g != null) this.fogBase.g = +c.g;
            if (c.b != null) this.fogBase.b = +c.b;
        }

        if (fog.distance != null) {
            const v = +fog.distance;
            if (Number.isFinite(v)) this.fogDistance = v;
        }

        if (fog.densityDay != null) {
            const v = +fog.densityDay;
            if (Number.isFinite(v)) this.fogDensityDay = v;
        }

        if (fog.densityNight != null) {
            const v = +fog.densityNight;
            if (Number.isFinite(v)) this.fogDensityNight = v;
        }

        if (ENGINE && ENGINE.log && ENGINE.log.debug) {
            ENGINE.log.debug(
                "[sky][fog] applyCfg color=(" + this.fogBase.r + "," + this.fogBase.g + "," + this.fogBase.b + ")" +
                " densityDay=" + this.fogDensityDay +
                " densityNight=" + this.fogDensityNight +
                " distance=" + this.fogDistance
            );
        }
    }

    init(render) {
        if (!render || typeof render.fogCfg !== "function") {
            throw new Error("[fog] render.fogCfg(cfg) is required");
        }

        render.fogCfg({
            color: [this.fogBase.r, this.fogBase.g, this.fogBase.b],
            density: this.fogDensityDay,
            distance: this.fogDistance
        });

        this._lastFogDensity = NaN;
        this._lastFogColorKey = "";
        this._lastFogDistance = NaN;

        if (ENGINE && ENGINE.log && ENGINE.log.debug) {
            ENGINE.log.debug("[sky][fog] init applied");
        }
    }

    update(render, celEval) {
        const d = (celEval && typeof celEval.dayFactor === "number") ? celEval.dayFactor : 1.0;

        const fogD = SkyMath.lerp(this.fogDensityNight, this.fogDensityDay, d);
        const key = SkyMath.rgbKey(this.fogBase.r, this.fogBase.g, this.fogBase.b);

        const changed =
            Math.abs(fogD - this._lastFogDensity) > 0.002 ||
            key !== this._lastFogColorKey ||
            Math.abs(this.fogDistance - this._lastFogDistance) > 0.1;

        if (!changed) return;

        if (ENGINE && ENGINE.log && ENGINE.log.debug) {
            ENGINE.log.debug(
                "[sky][fog] APPLY density=" + fogD.toFixed(4) +
                " distance=" + this.fogDistance.toFixed(1) +
                " dayFactor=" + d.toFixed(4)
            );
        }

        render.fogCfg({
            color: [this.fogBase.r, this.fogBase.g, this.fogBase.b],
            density: fogD,
            distance: this.fogDistance
        });

        this._lastFogDensity = fogD;
        this._lastFogColorKey = key;
        this._lastFogDistance = this.fogDistance;
    }

    destroy() {
    }
}

module.exports = FogController;

File: assets/Scripts/environment/sky/FogController.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
local LuaNumber = luaRuntime.LuaNumber
local LuaNumberIsFinite = luaRuntime.LuaNumberIsFinite
local LuaClass = luaRuntime.LuaClass
local LuaNumberToFixed = luaRuntime.LuaNumberToFixed
local LuaArrayJoin = luaRuntime.LuaArrayJoin
function req(self, v, msg)
    if v == nil then
        error(
            LuaConstruct(Error, msg),
            0
        )
    end
    return v
end
function clamp(self, x, a, b)
    x = LuaNumber(x)
    if not LuaNumberIsFinite(x) then
        return a
    end
    local lua_temp_1
    if x < a then
        lua_temp_1 = a
    else
        local lua_temp_0
        if x > b then
            lua_temp_0 = b
        else
            lua_temp_0 = x
        end
        lua_temp_1 = lua_temp_0
    end
    return lua_temp_1
end
function lerp(self, a, b, t)
    return a + (b - a) * t
end
FogController = LuaClass()
FogController.name = "FogController"
function FogController.prototype.lua_constructor(self)
    self.day = {
        r = 0.7,
        g = 0.78,
        b = 0.9,
        density = 0.006,
        distance = 260
    }
    self.night = {
        r = 0.06,
        g = 0.08,
        b = 0.12,
        density = 0.01,
        distance = 140
    }
    self._cfgRef = nil
    self._render = nil
    self._lastKey = ""
end
function FogController.prototype.init(self, render)
    self._render = req(_G, render, "[sky][fog] render is required")
end
function FogController.prototype.destroy(self)
end
function FogController.prototype.applyCfg(self, cfg)
    if cfg == self._cfgRef then
        return
    end
    self._cfgRef = cfg
    if cfg.fogDay then
        local f = cfg.fogDay
        if f.r ~= nil then
            self.day.r = LuaNumber(f.r)
        end
        if f.g ~= nil then
            self.day.g = LuaNumber(f.g)
        end
        if f.b ~= nil then
            self.day.b = LuaNumber(f.b)
        end
        if f.density ~= nil then
            self.day.density = LuaNumber(f.density)
        end
        if f.distance ~= nil then
            self.day.distance = LuaNumber(f.distance)
        end
    end
    if cfg.fogNight then
        local f = cfg.fogNight
        if f.r ~= nil then
            self.night.r = LuaNumber(f.r)
        end
        if f.g ~= nil then
            self.night.g = LuaNumber(f.g)
        end
        if f.b ~= nil then
            self.night.b = LuaNumber(f.b)
        end
        if f.density ~= nil then
            self.night.density = LuaNumber(f.density)
        end
        if f.distance ~= nil then
            self.night.distance = LuaNumber(f.distance)
        end
    end
end
function FogController.prototype.update(self, render, cel)
    render = req(_G, render, "[sky][fog] render is required")
    req(
        _G,
        cel and KTypeOf(cel.dayFactor) == "number",
        "[sky][fog] cel.dayFactor is required"
    )
    local df = clamp(_G, cel.dayFactor, 0, 1)
    local r = lerp(_G, self.night.r, self.day.r, df)
    local g = lerp(_G, self.night.g, self.day.g, df)
    local b = lerp(_G, self.night.b, self.day.b, df)
    local density = clamp(
        _G,
        lerp(_G, self.night.density, self.day.density, df),
        0,
        0.03
    )
    local distance = math.max(
        25,
        lerp(_G, self.night.distance, self.day.distance, df)
    )
    local key = LuaArrayJoin(
        {
            LuaNumberToFixed(r, 4),
            LuaNumberToFixed(g, 4),
            LuaNumberToFixed(b, 4),
            LuaNumberToFixed(density, 6),
            LuaNumberToFixed(distance, 2)
        },
        "|"
    )
    if key == self._lastKey then
        return
    end
    self._lastKey = key
    render:fogCfg({color = {r = r, g = g, b = b}, density = density, distance = distance})
end
M = FogController

return M
